### DISCUSSION TITLE
Displaying inline plots with matplotlib

### DISCUSSION THREAD

**Discussion by icweaver** [cjdoris/PythonCall.jl#86] [Q&A] [answered]
Hi,

Thanks for creating such a wonderful package, I've been having a blast playing around with it. I was wondering how I might go about displaying plots produced by `matplotlib` directly in a Pluto notebook, instead of directing it to an interactive gui? I see that this happens automatically for your [`seaborn`](https://github.com/cjdoris/PythonCall.jl#example-1-calling-python-from-julia) example, but trying a simpler example directly with `matplotlib` gives me the following:

* Initial set-up in fresh directory
```
> julia --project=@.
pkg> add PythonCall, CondaPkg, Pluto
julia> using CondaPkg
julia> CondaPkg.add("matplotlib")
julia> using Pluto
julia> Pluto.run()
```

* Pluto notebook in same directory
```julia
begin
    import Pkg
    Pkg.activate(Base.current_project())

    using PythonCall
end
```

```julia
@py begin
	import matplotlib as mpl
	import matplotlib.pyplot as plt

	mpl.use("agg")
end
```

```julia
p = plt.plot(rand(10))
# p = Python list: [<matplotlib.lines.Line2D object at 0x7fcd403bee90>]
```

```julia
plt.show()
# Python None
```
```
sys:1: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
```

* System info
```
Julia Version 1.7.0
Commit 3bf9d17731 (2021-11-30 12:12 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, tigerlake)
```

Trying `display` as described [here](https://cjdoris.github.io/PythonCall.jl/dev/compat/#MatPlotLib-/-PyPlot-/-Seaborn) also gives me similar issues, but maybe I am misusing them here?

```julia
display(p) # returns/displays nothing, regardless of backend
```

Using the default interactive backend (`QtAgg` on my system) works just fine, but I would really love to be able to have the plots show up directly in the notebook if possible

At any rate, happy holidays and thanks again for providing this package!

**Comment by cjdoris** [comment, marked as answer]
Happy holidays!

PythonCall knows how to display a `matplotlib.Figure` so you need to return one of those from a Pluto cell.

For example something like
```julia
plt.plot(.....)
plt.gcf()
```
or
```julia
fig = plt.figure()
fig.plot(.....)
fig
```
should work.

**Reply by icweaver** [reply]
Hi, thanks for getting back to me! So I think I am misunderstanding your suggestions here. Both examples (did you also mean to put `plt.plot` instead of `fig.plot` for the second one?) return the same error for me:

![Screenshot from 2022-01-03 17-32-16](https://user-images.githubusercontent.com/25312320/147987578-a1ae5e87-31b1-44ee-981e-00c5c5d7e038.png)
(Pluto.jl v0.17.4, PythonCall.jl v0.5.0, CondaPkg.jl v0.2.3) 

Is there another way that I should be importing/using `matplotlib` for this example?

**Reply by cjdoris** [reply]
Ah, looks like a bug in the multimedia display. I'll fix it when I'm back at my computer in the next few days.

**Reply by icweaver** [reply]
Sounds great, thanks!

**Reply by cjdoris** [reply]
I've fixed this on the main branch now.

**Comment by icweaver** [comment]
This is amazing, thanks so much!! Being able to go between inline/gui and julia/python objects so seamlessly feels like witchcraft 🪄

![Screenshot from 2022-01-07 14-36-48](https://user-images.githubusercontent.com/25312320/148597758-9b4d325a-9a3f-456b-9fb8-38e6a55cdc98.png)

**Comment by icweaver** [comment]
Do you have a feel for when the next release may be? I thought it would be great to share Pluto notebooks [like this one](https://icweaver.github.io/WASP-50b/html/fun_with_python.jl.html) with colleagues to use Pluto's [built-in package manager](https://github.com/fonsp/Pluto.jl/wiki/%F0%9F%8E%81-Package-management) to automatically download and setup Python on their system for them. Currently, the manager can only pull down released packages I think, so the plot under the "Using packages" section cannot be shown just yet. Totally understand if you are trying to add in other things first before tagging the next release though!

**Reply by cjdoris** [reply]
Sure, I've just submitted v0.5.1 for registration.
